qcfractal/queue/handlers.py

```diff
@@ -65,7 +68,46 @@ def put(self):
             raise tornado.web.HTTPError(status_code=400, reason="Id or ResultId must be specified.")
 
         if body.meta.operation == "restart":
-            tasks_updated = self.storage.queue_reset_status(**body.data.dict(), reset_error=True)
+            d = body.data.dict()
+            d.pop("new_tag", None)
+            d.pop("new_priority", None)
+            tasks_updated = self.storage.queue_reset_status(**d, reset_error=True)
+            data = {"n_updated": tasks_updated}
+        elif body.meta.operation == "regenerate":
+            tasks_updated = 0
+            result_data = self.storage.get_procedures(id=body.data.base_result)["data"]
+
+            new_tag = body.data.new_tag
+            if body.data.new_priority is None:
+                new_priority = PriorityEnum.NORMAL
+            else:
+                new_priority = PriorityEnum(int(body.data.new_priority))
+
+            for r in result_data:
+                model = build_procedure(r)
+
+                # Only regenerate the task if the base record is not complete
```

Comment:
Too much application logic in this PUT method, the REST should not handle all that knowledge about the Queue and it's status. At most it can use one variable to redirect to different socket functions. I suggest moving that logic to other layers

Reply:
Completely agree. However, was trying to minimize changes to that file due to your REST rewriting

Reply:
This file is fully refactored already in the REST refactor too :)

Reply:
Not the socket, which would be too low level. There isn't a very good place for it yes (except maybe procedures?) but we are definitely planning on revisiting this.
handlers.py is being modified in another refactor at the moment so it might be best to just leave it for the time being.

Reply:
Absolutely agree. It has to live somewhere, and I think it's fine where it rests for now.